plugin/storage/grpc/config/config.go

```diff
@@ -65,13 +65,36 @@ func (c *Configuration) Build() (shared.StoragePlugin, error) {
 
 	storagePlugin, ok := raw.(shared.StoragePlugin)
 	if !ok {
-		return nil, fmt.Errorf("unexpected type for plugin \"%s\"", shared.StoragePluginIdentifier)
+		return nil, fmt.Errorf("unable to cast %T to shared.StoragePlugin for plugin \"%s\"",
+			raw, shared.StoragePluginIdentifier)
 	}
+	archiveStoragePlugin, ok := raw.(shared.ArchiveStoragePlugin)
```

Comment:
This looks like we came back to my original proposal of doing a runtime check to determine whether the plugin implements additional interfaces. However, I am puzzled by how this works, given that `raw` is an `interface{}` proxy that's apparently created by the plugin framework to represent the attached plugin binary (which technically may not be even in Go). How does the framework determine that the proxy need to implement these internal interfaces?

Reply:
`raw` is type of `shared.grpcClient` which implements `shared.ArchiveStoragePlugin` and `shared.PluginCapabilities`. Jaeger will call `Capabilities()` method which have following error handling: https://github.com/jaegertracing/jaeger/pull/2317/files#diff-4ad2ef46d844069ff3a195fb9a785ed4R219. If plugin doesn't have `Capabilities` method we fallback to `archive storage not supported` logic.